### PR TITLE
Rework new-app

### PR DESCRIPTION
**Pull request by kargakis**

### CODE REVIEW

File: pkg/generate/app/componentref.go [hunk 1]
```diff
@@ -42,20 +44,22 @@ func componentWithSource(s string) (component, repo string, builder bool, err er
 	} else {
 		component = s
 	}
-	// TODO: component must be of the form compatible with a pull spec *or* <namespace>/<name>
+	// component must be of the form compatible with a pull spec *or* <namespace>/<name>
```
Comment (kargakis) [inline pkg/generate/app/componentref.go]: @smarterclayton removed this TODO since `ParseDockerImageReference` already checks for both cases, [ain't it](https://github.com/openshift/origin/blob/e6d024bc78fc682c081aba1a461bb57bc8577f77/pkg/image/api/helper.go#L44)?

Comment (kargakis) [inline pkg/generate/app/componentref.go]: @smarterclayton also can you give me an example of an [image]~[code] form?

**Comment by kargakis** [comment]
@smarterclayton also an example of [this](https://github.com/openshift/origin/blob/e3e4b61fe27d303bef36a8b5414fd71accbe8ea1/pkg/generate/app/componentref.go#L245) input?

**Comment by kargakis** [comment]
> @smarterclayton also an example of this input?

Got it about the plus sign. We use it to group components.

**Comment by smarterclayton** [comment]
Please make sure your changes and @fabianofranz do not overlap - he has a lot of important changes going in here and I don't want either of you to be blocked on the other.  Will review later.

**Comment by fabianofranz** [comment]
Here is my PR: https://github.com/openshift/origin/pull/1716. I'm working to remove WIP and be ready to review later today.

**Comment by kargakis** [comment]
> Here is my PR: #1716. I'm working to remove WIP and be ready to review later today.

Cool. I am just getting into the generate package and new-app code so it will be a while before I finish this.

File: pkg/generate/app/pipeline.go [hunk 2]
```diff
@@ -180,43 +181,62 @@ func (s sortablePorts) Swap(i, j int) {
...
 func AddServices(objects Objects) Objects {
 	svcs := []runtime.Object{}
 	for _, o := range objects {
 		switch t := o.(type) {
 		case *deploy.DeploymentConfig:
 			for _, container := range t.Template.ControllerTemplate.Template.Spec.Containers {
+				name, generateName := makeValidServiceName(t.Name)
+				svc := &kapi.Service{
+					ObjectMeta: kapi.ObjectMeta{
+						Name:         name,
```
Comment (kargakis) [inline pkg/generate/app/pipeline.go]: @smarterclayton every service gets its name from the name of the deploymentConfig. But what if that deployment specifies more than one container in a pod? As is, say there is a pod with three containers, we will get three services with the same name.

File: pkg/generate/app/pipeline.go [hunk 3]
```diff
@@ -180,43 +181,62 @@ func (s sortablePorts) Swap(i, j int) {
...
+// AddRoutes sets up routes for the provided objects
+func AddRoutes(objects Objects) Objects {
+	routes := []runtime.Object{}
+	for _, o := range objects {
+		switch t := o.(type) {
+		case *kapi.Service:
+			for range t.Spec.Ports {
+				routes = append(routes, &route.Route{
+					ObjectMeta: kapi.ObjectMeta{
+						Name:   t.Name,
```
Comment (kargakis) [inline pkg/generate/app/pipeline.go]: @smarterclayton here as well, every route gets its name from the service it will connect to. But what happens with a multiport service? Every route on it will have the same name as is.

Comment (smarterclayton) [inline pkg/generate/app/pipeline.go]: Routes will have to have a port mapping, but routers decide what outbound ports you get (@pweil-).  I.e., a user can say "I want to expose port 443 from containers via the router", but only the router knows what port that actually is.

----- Original Message -----

Comment (smarterclayton) [inline pkg/generate/app/pipeline.go]: I suspect routes should declare named bindings for protocols:

"http": 8080,  # container port
"https": 8443, # container port

and the router decides what to do with "http" (it either supports it or not) and then the router scheduler will update the status of the route to say what actual port and IP and DNS address that port gets exposed on.

I.e.

Pod exposes 8080, 8443
Service exposes 8443
Route points to service, defines "http":8080, "https":8443
Route Allocator binds Route to Router Foo
Router Foo listens on 80 for "http" and 443 for "https", so the status of the route gets set to:

"status": {
  "router": "foo",
  "bindings": [{
    "name": "http",
    "port": "80",
    "dns": "route-namespace.myrouter.west.com"
  }]
}

None of that is there today, needs more thought and design.

----- Original Message -----

**Comment by kargakis** [comment]
How I currently output multiport services:
```
Service "mysql" created at 172.30.202.237:3306,172.30.202.237:8080 to talk to pods over port(s) 3306,8080.
```

**Comment by kargakis** [comment]
(Multi)port overriding:
```
[vagrant@openshiftdev sample-app]$ osc new-app https://github.com/openshift/ruby-hello-world mysql -l name=hello-world --port=40 --port=5
deploymentConfigs/mysql
imageStreams/ruby-hello-world
buildConfigs/ruby-hello-world
services/mysql
routes/mysql
A build was created - you can run `osc start-build ruby-hello-world` to start it.
Service "mysql" created at 172.30.2.169:40,172.30.2.169:5 to talk to pods over port(s) 40,5.
```

**Comment by kargakis** [comment]
Name overriding:
```
[vagrant@openshiftdev sample-app]$ osc new-app https://github.com/openshift/ruby-hello-world mysql --name=initial -l name=hello-world
deploymentConfigs/initial
imageStreams/initial
buildConfigs/initial
services/initial
routes/initial
A build was created - you can run `osc start-build initial` to start it.
Service "initial" created at 172.30.235.75:3306,172.30.235.75:8080 to talk to pods over port(s) 3306,8080.
```

**Comment by kargakis** [comment]
Can I get a review on this? @smarterclayton @bparees @csrwng

**Comment by smarterclayton** [comment]
This seems confusing to users, since there's no mapping possible.  Why do we need this?

**Comment by kargakis** [comment]
> This seems confusing to users, since there's no mapping possible. Why do we need this?

We already let the user specify a port in `ex generate` and since we are dealing with multiport services here I imagined we would want this. If that's not the case, I'll switch it back to specifying only one port at most.

**Comment by bparees** [comment]
for port overriding (and for that matter name overriding) I think we need a scheme that allows you to override the ports+name associated with each "input object" into new-app.  that is, I should be able to run:

osc new-app mysql@1024=userDB postgresql@2048=inventoryDB

and end up with a mysql service named "userDB" running on port 1024 and  postgres service named "inventoryDB" running on 2048.  Both the service and the deploymentconfig would share the name.

(that particular syntax is just an example, but hopefully you take my point).

the challenge is what to do when the "input object" is a template.  In that case it's probably not valid to override names or ports.

but I should be able to do:   
osc new-app http://github.com/someguy/someproject.git@1234=myFrontend  mysql@3306=myBackend

Again, don't take the syntax as gospel, but that's the concept.

@smarterclayton thoughts?  Obviously we need to lay out all the characters we're using today, I know new-app allows some pairing syntax with "," and "~" that probably make my proposal more complicated, but whatever we come up with I think it has to have that level of flexibility.

**Comment by smarterclayton** [comment]
I don't want to support that syntax yet, although eventually it might be necessary.  In the short term that is what 'osc expose' / 'kubectl expose' does so we don't need to duplicate the function here.

----- Original Message -----

**Comment by kargakis** [comment]
Hmm, a scheme would be the best solution to this and I am inclined not to override ports at all for the time being.

File: pkg/generate/app/pipeline.go [hunk 4]
```diff
@@ -33,8 +34,6 @@ func NewImagePipeline(from string, image *ImageRef) (*Pipeline, error) {
 }
 
 func NewBuildPipeline(from string, input *ImageRef, strategy *BuildStrategyRef, source *SourceRef) (*Pipeline, error) {
-	strategy.Base = input
```
Comment (kargakis) [inline pkg/generate/app/pipeline.go]: The base image setup is now handled in `StrategyAndSourceForRepository` by `FromSourceRef`.

File: hack/test-cmd.sh [hunk 5]
```diff
@@ -333,13 +333,13 @@ osc delete imageRepositories test
...
 [ -n "$(osc get imageRepositories mysql -t "{{ index .metadata.annotations \"openshift.io/image.dockerRepositoryCheck\"}}")" ]
-[ "$(osc new-app php mysql -o yaml | grep 3306)" ]
+[ "$(osc new-app mysql -o yaml | grep 3306)" ]
 # verify we can generate a Docker image based component "mongodb" directly
 [ ! "$(osc new-app unknownhubimage -o yaml)" ]
 [ "$(osc new-app mongo -o yaml | grep library/mongo)" ]
 # the local image repository takes precedence over the Docker Hub "mysql" image
 [ "$(osc new-app mysql -o yaml | grep mysql-55-centos7)" ]
-osc new-app php mysql
+osc new-app mysql
```
Comment (smarterclayton) [inline hack/test-cmd.sh]: Why did you remove this test?

File: hack/test-cmd.sh [hunk 6]
```diff
@@ -350,6 +350,12 @@ osc get template ruby-helloworld-sample
 # create from template with code explicitly set is not supported
 [ ! "$(osc new-app ruby-helloworld-sample~git@github.com/mfojtik/sinatra-app-example)" ]
 osc delete template ruby-helloworld-sample
+# override component names
+[ "$(osc new-app mysql --name=db | grep db)" ]
+# see if we get any route
+osc new-app https://github.com/openshift/ruby-hello-world -l name=ruby
+[ "$(osc describe route ruby-hello-world | grep name=ruby)" ]
```
Comment (smarterclayton) [inline hack/test-cmd.sh]: Use app=ruby instead.  We need to stop using "name" as a label.

File: pkg/cmd/cli/cmd/newapp.go [hunk 7]
```diff
@@ -169,8 +171,12 @@ func RunNewApplication(f *clientcmd.Factory, out io.Writer, c *cobra.Command, ar
 	for _, item := range result.List.Items {
 		switch t := item.(type) {
 		case *kapi.Service:
-			// TODO: handle multi-port created services
-			fmt.Fprintf(c.Out(), "Service %q created at %s:%d to talk to pods over port %d.\n", t.Name, t.Spec.PortalIP, t.Spec.Ports[0].Port, t.Spec.Ports[0].TargetPort.IntVal)
+			var portals, targetPorts []string
+			for _, port := range t.Spec.Ports {
+				portals = append(portals, fmt.Sprintf("%s:%d", t.Spec.PortalIP, port.Port))
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: I'd prefer to show port mappings if they exist.  See describeServicePorts in projectstatus.go for the preferred output form.

File: pkg/generate/app/cmd/newapp.go [hunk 8]
```diff
@@ -325,10 +337,17 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
 				if err != nil {
 					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
 				}
-				strategy, source, err := app.StrategyAndSourceForRepository(ref.Input().Uses)
+				strategy, source, err := app.StrategyAndSourceForRepository(ref.Input().Uses, app.PerfectMatchWeightedResolver{
```
Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: Why have a different resolver here?

File: pkg/generate/app/cmd/newapp.go [hunk 9]
```diff
@@ -451,6 +470,7 @@ func (c *AppConfig) Run(out io.Writer) (*AppResult, error) {
 	}
 
 	objects = app.AddServices(objects)
+	objects = app.AddRoutes(objects)
```
Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: Pretty sure we said we wouldn't add routes automatically.

File: pkg/generate/app/componentref.go [hunk 10]
```diff
@@ -191,6 +222,7 @@ func (r PerfectMatchWeightedResolver) Resolve(value string) (*ComponentMatch, er
 	case 0:
 		return nil, ErrNoMatch{value: value}
 	case 1:
+		// This is an inexact match so our PerfectMatchWeightedResolver is not returning only perfect matches...
```
Comment (smarterclayton) [inline pkg/generate/app/componentref.go]: You can be inexact but still perfect.  DockerHub images are inexact, but perfect.

File: pkg/generate/app/cmd/newapp.go (hunk 8 above)
Comment (kargakis) [inline pkg/generate/app/cmd/newapp.go]: It's not different, I am just merging the docker and imageStream resolvers from AppConfig into one.

File: hack/test-cmd.sh (hunk 5 above)
Comment (kargakis) [inline hack/test-cmd.sh]: Because it's PHP. No, seriously, it failed in tests because the php image probably lacks some configuration regarding its ports. I'll have to run Travis again with it in order to remember the exact error message.

File: pkg/generate/app/cmd/newapp.go (hunk 9 above)
Comment (kargakis) [inline pkg/generate/app/cmd/newapp.go]: Hmmm, I don't remember that. Anyway, how should it be done then? Another flag to enable it?

Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: Just don't add routes for now.

----- Original Message -----

File: hack/test-cmd.sh (hunk 5 above)
Comment (smarterclayton) [inline hack/test-cmd.sh]: I'd prefer we find out how to keep it.

----- Original Message -----

File: pkg/generate/app/cmd/newapp.go (hunk 8 above)
Comment (kargakis) [inline pkg/generate/app/cmd/newapp.go]: Oh yea, I have changed the weight in the docker resolver. Well, I am not sure about it to be honest. @csrwng is doing the same in [generate](https://github.com/openshift/origin/blob/09f87e4d4e975d96948900ed6a8edee4365bc380/pkg/cmd/experimental/generate/generate.go#L218)

Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: We should only need to create one resolver for the entire appconfig - the initializer up above that creates it should ensure this call gets the same resolver.

Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: Cesar why did we change the weight?  Or was it just copied *before* new-app existed.

Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: We probably shouldn't be doing resolution so late.  Why is that necessary?

File: pkg/generate/app/sourcelookup.go [hunk 11]
```diff
@@ -215,18 +234,37 @@ func (e SourceRepositoryEnumerator) Detect(dir string) (*SourceRepositoryInfo, e
...
+// of the provided source repository
+func StrategyAndSourceForRepository(repo *SourceRepository, resolver Resolver) (*BuildStrategyRef, *SourceRef, error) {
+	srcRef, err := NewSourceRefGenerator().FromGitURL(repo.location)
 	if err != nil {
-		return nil, nil, fmt.Errorf("cannot obtain remote URL for repository at %s", repo.location)
+		return nil, nil, fmt.Errorf("cannot obtain remote URL for repository at %s: %v", repo.location, err)
 	}
-	strategy := &BuildStrategyRef{
-		IsDockerBuild: repo.IsDockerBuild(),
+	strategy, err := NewBuildStrategyRefGenerator(source.DefaultDetectors, resolver).FromSourceRef(srcRef)
```
Comment (smarterclayton) [inline pkg/generate/app/sourcelookup.go]: I don't like doing resolution in these methods.  All resolution should happen before this method gets called (specifically for new-app) at the very early part of the flow.

File: hack/test-cmd.sh (hunk 5 above)
Comment (kargakis) [inline hack/test-cmd.sh]: Here's the error: 
```
Error: Service "php" is invalid: spec.ports: required value
```

Comment (smarterclayton) [inline hack/test-cmd.sh]: We shouldn't generate a service for PHP if it exposes no ports.

----- Original Message -----

**Comment by kargakis** [comment]
@smarterclayton re ` osc new-app php mysql` on current master:

```
services/mysql
deploymentConfigs/php
deploymentConfigs/mysql
Service "mysql" created at 172.30.199.98:3306 to talk to pods over port 3306.
```

This does not create any buildConfig and imageStream, so by setting up build triggers I get the current error on Travis:
```
E0427 10:14:45.467695    7735 factory.go:68] couldn't create initial deployment for config default/mysql-55-centos7:0: DeploymentConfig "mysql-55-centos7" is invalid: triggers[1].imageChange.from: invalid value 'mysql-55-centos7': imageStream "mysql-55-centos7" not found
```

Obvious since there is no stream. I guess we should inspect if there are no component passed to new-app that are expected to be built and not *try* to create buildConfigs and streams?

**Comment by kargakis** [comment]
Actually does it make sense for new-app to set up just services?

**Comment by smarterclayton** [comment]
It does, since the service provides you grouping and the linking of the elements of the app together.

**Comment by smarterclayton** [comment]
That looks like a regression.  An image stream should be created for external images we reference.  However, where is mysql-55-centos7 coming from?  If that's an existing image stream (and new-app detected it) we should be pointing to that image stream in the triggers (and it will be across namespaces).

----- Original Message -----

**Comment by smarterclayton** [comment]
What's left on this?

**Comment by kargakis** [comment]
> What's left on this?

I just want to fix this last test on `Run` tomorrow and it's done.

**Comment by kargakis** [comment]
[test]

**Comment by kargakis** [comment]
@smarterclayton this is ok now.

**Comment by smarterclayton** [comment]
LGTM [merge]
